### PR TITLE
 PROCESSINGS: Retrieve All External Countries

### LINKED ISSUE #7: PROCESSINGS: Retrieve All External Countries

### PR DESCRIPTION
**Pull request by Mehdi-Aghaei**
Closes #7

### CODE REVIEW

**Comment by Mehdi-Aghaei** [comment]
@hassanhabib 
hi Brother, can you take look at this Pr is it enough to balance architecture?
also, I finish the upsert and already merged if it's possible take a look at that too.

File: CoronaTracker.Core.Tests.Unit/Services/Processings/ExternalCountries/ExternalCountryProcessingServiceTests.Exceptions.RetrieveAll.cs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+            this.loggingBrokerMock.Verify(broker =>
+                broker.LogError(It.Is(SameExceptionAs(
+                    expectedExternalCountryProcessingDependencyException))),
+                        Times.Once);
+
+            this.externalCountryServiceMock.VerifyNoOtherCalls();
+            this.loggingBrokerMock.VerifyNoOtherCalls();
+        }
+        [Fact]
```
Comment (hassanhabib) [inline CoronaTracker.Core.Tests.Unit/Services/Processings/ExternalCountries/ExternalCountryProcessingServiceTests.Exceptions.RetrieveAll.cs:56]: Need a new line to separate these two methods.

File: CoronaTracker.Core/Services/Processings/ExternalCountries/ExternalCountryProcessingService.Exceptions.cs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+            catch (ExternalCountryDependencyException externalCountryDependencyException)
+            {
+
+                throw CreateAndLogDependencyException(externalCountryDependencyException);
+            }     
+            catch (ExternalCountryServiceException externalCountryServiceException)
+            {
+
+                throw CreateAndLogDependencyException(externalCountryServiceException);
+            }catch(Exception exception)
```
Comment (hassanhabib) [inline CoronaTracker.Core/Services/Processings/ExternalCountries/ExternalCountryProcessingService.Exceptions.cs]: Run code clean up across the whole solution

File: CoronaTracker.Core/Services/Processings/ExternalCountries/ExternalCountryProcessingService.cs [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+            this.externalCountryService = externalCountryService;
+            this.loggingBroker = loggingBroker;
+        }
+
+        public ValueTask<List<ExternalCountry>> RetrieveAllExternalCountriesAsync() =>
+        TryCatch(async () => 
+        {
+            return await this.externalCountryService.RetrieveAllExternalCountriesAsync();
+        });    
+
```
Comment (hassanhabib) [inline CoronaTracker.Core/Services/Processings/ExternalCountries/ExternalCountryProcessingService.cs:32]: Remove this empty line

File: CoronaTracker.Core/Services/Processings/ExternalCountries/ExternalCountryProcessingService.cs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        {
+            this.externalCountryService = externalCountryService;
+            this.loggingBroker = loggingBroker;
+        }
+
+        public ValueTask<List<ExternalCountry>> RetrieveAllExternalCountriesAsync() =>
+        TryCatch(async () => 
+        {
+            return await this.externalCountryService.RetrieveAllExternalCountriesAsync();
+        });
```
Comment (hassanhabib) [inline CoronaTracker.Core/Services/Processings/ExternalCountries/ExternalCountryProcessingService.cs:31]: We could probably one-liner this guy?

File: CoronaTracker.Core/Models/Processings/ExternalCountries/ExternalCountryProcessingDependencyException.cs [hunk 5]
```diff
@@ -0,0 +1,16 @@
+﻿// ---------------------------------------------------------------
```
Comment (ShriHumrudha) [inline CoronaTracker.Core/Models/Processings/ExternalCountries/ExternalCountryProcessingDependencyException.cs:1]: The exceptions should be nested under 'Exceptions' folder

File: CoronaTracker.Core/Models/Processings/ExternalCountries/ExternalCountryProcessingServiceException.cs [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+// ---------------------------------------------------------------
+
+using Xeptions;
+
+namespace CoronaTracker.Core.Models.Processings.ExternalCountries.Exceptions
+{
+    public class ExternalCountryProcessingServiceException : Xeption
+    {
+        public ExternalCountryProcessingServiceException(Xeption innerException)
+            : base("External country processing service error occurred, please contact support", innerException)
```
Comment (ShriHumrudha) [inline CoronaTracker.Core/Models/Processings/ExternalCountries/ExternalCountryProcessingServiceException.cs:13]: Parameter Aliasing, Remove processing word
Take a look in other files as well

File: CoronaTracker.Core/Services/Processings/ExternalCountries/ExternalCountryProcessingService.Exceptions.cs [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+using Xeptions;
+
+namespace CoronaTracker.Core.Services.Processings.ExternalCountries
+{
+    public partial class ExternalCountryProcessingService
+    {
+        private delegate ValueTask<List<ExternalCountry>> ReturningExternalCountriesFunction();
+
+        private async ValueTask<List<ExternalCountry>>
+        TryCatch(ReturningExternalCountriesFunction returningExternalCountriesFunction)
```
Comment (ShriHumrudha) [inline CoronaTracker.Core/Services/Processings/ExternalCountries/ExternalCountryProcessingService.Exceptions.cs:22]: Suggestion: Move TryCatch above and the parameter below.

File: CoronaTracker.Core/Services/Processings/ExternalCountries/ExternalCountryProcessingService.Exceptions.cs [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+            }catch(Exception exception)
+            {
+                var failedExternalCountryProcessingServiceException =
+                    new FailedExternalCountryProcessingServiceException(exception);
+
+                throw CreateAndLogServiceException(failedExternalCountryProcessingServiceException);
+            }
+        }
+
+        private ExternalCountryProcessingServiceException CreateAndLogServiceException(Xeption exception)
```
Comment (ShriHumrudha) [inline CoronaTracker.Core/Services/Processings/ExternalCountries/ExternalCountryProcessingService.Exceptions.cs:48]: Suggestion: Move this method to the last.

File: CoronaTracker.Core.Tests.Unit/Services/Processings/ExternalCountries/ExternalCountryProcessingServiceTests.Exceptions.RetrieveAll.cs [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+{
+    public partial class ExternalCountryProcessingServiceTests
+    {
+        [Theory]
+        [MemberData(nameof(DependencyExceptions))]
+        public async Task ShouldThrowDependencyExceptionOnRetrieveAllIfDependencyErrorOccursAndLogItAsync(
+            Xeption dependencyException)
+        {
+            // given 
+            var someCountries = CreateRandomExternalCountries();
```
Comment (ShriHumrudha) [inline CoronaTracker.Core.Tests.Unit/Services/Processings/ExternalCountries/ExternalCountryProcessingServiceTests.Exceptions.RetrieveAll.cs:26]: Explicit left-hand side

File: CoronaTracker.Core.Tests.Unit/Services/Processings/ExternalCountries/ExternalCountryProcessingServiceTests.Exceptions.RetrieveAll.cs [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+
+            var expectedExternalCountryProcessingDependencyException =
+                new ExternalCountryProcessingDependencyException(
+                    dependencyException.InnerException as Xeption);
+
+            this.externalCountryServiceMock.Setup(service =>
+                service.RetrieveAllExternalCountriesAsync())
+                .ThrowsAsync(dependencyException);
```
Comment (ShriHumrudha) [inline CoronaTracker.Core.Tests.Unit/Services/Processings/ExternalCountries/ExternalCountryProcessingServiceTests.Exceptions.RetrieveAll.cs:34]: One more tab indendation

File: CoronaTracker.Core.Tests.Unit/Services/Processings/ExternalCountries/ExternalCountryProcessingServiceTests.Exceptions.RetrieveAll.cs [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+        {
+            // given
+            var someCountries = CreateRandomExternalCountries();
+
+            var serviceException = new Exception();
+
+            var failedExternalCountryProcessingServiceException =
+                new FailedExternalCountryProcessingServiceException(serviceException);
+
+            var expectedCountryProcessingServiveException =
```
Comment (ShriHumrudha) [inline CoronaTracker.Core.Tests.Unit/Services/Processings/ExternalCountries/ExternalCountryProcessingServiceTests.Exceptions.RetrieveAll.cs:67]: expectedExternalCountryProcessingServiveException ??

File: CoronaTracker.Core.Tests.Unit/Services/Processings/ExternalCountries/ExternalCountryProcessingServiceTests.Exceptions.RetrieveAll.cs [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+                new ExternalCountryProcessingServiceException(
+                    failedExternalCountryProcessingServiceException);
+
+            this.externalCountryServiceMock.Setup(service =>
+                service.RetrieveAllExternalCountriesAsync())
+                    .Throws(serviceException);
+
+            // when
+            ValueTask<List<ExternalCountry>> upsertCountryTask =
```
Comment (ShriHumrudha) [inline CoronaTracker.Core.Tests.Unit/Services/Processings/ExternalCountries/ExternalCountryProcessingServiceTests.Exceptions.RetrieveAll.cs:76]: Change variable name

File: CoronaTracker.Core.Tests.Unit/Services/Processings/ExternalCountries/ExternalCountryProcessingServiceTests.cs [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+    {
+        private readonly Mock<IExternalCountryService> externalCountryServiceMock;
+        private readonly Mock<ILoggingBroker> loggingBrokerMock;
+        private readonly IExternalCountryProcessingService externalCountryProcessingService;
+
+        public ExternalCountryProcessingServiceTests()
+        {
+            this.externalCountryServiceMock = new Mock<IExternalCountryService>();
+            this.loggingBrokerMock = new Mock<ILoggingBroker>();
+            this.externalCountryProcessingService = new ExternalCountryProcessingService(
```
Comment (ShriHumrudha) [inline CoronaTracker.Core.Tests.Unit/Services/Processings/ExternalCountries/ExternalCountryProcessingServiceTests.cs:32]: New line break.

File: CoronaTracker.Core.Tests.Unit/Services/Processings/ExternalCountries/ExternalCountryProcessingServiceTests.cs [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+            return new TheoryData<Xeption>
+            {
+                new ExternalCountryDependencyException(innerException),
+                new ExternalCountryServiceException(innerException)
+            };
+        }
+
+        private static string GetRandomMessage() =>
+            new MnemonicString(wordCount: GetRandomNumber()).GetValue();
+        private List<ExternalCountry> CreateRandomExternalCountries() =>
```
Comment (ShriHumrudha) [inline CoronaTracker.Core.Tests.Unit/Services/Processings/ExternalCountries/ExternalCountryProcessingServiceTests.cs:53]: New line space

File: CoronaTracker.Core.Tests.Unit/Services/Processings/ExternalCountries/ExternalCountryProcessingServiceTests.cs [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+            return actualException =>
+                actualException.Message == expectedException.Message
+                && actualException.InnerException.Message == expectedException.InnerException.Message
+                && (actualException.InnerException as Xeption).DataEquals(expectedException.Data);
+        }
+
+        private static int GetRandomNumber() =>
+            new IntRange(min: 2, max: 10).GetValue();
+
+        private static Filler<ExternalCountry> CreateExternalCountryFiller()
```
Comment (ShriHumrudha) [inline CoronaTracker.Core.Tests.Unit/Services/Processings/ExternalCountries/ExternalCountryProcessingServiceTests.cs:66]: Single line code?

**Pull request merged**